File: java_version/Linked_List_Array/Partition_list.java
package Linked_List_Array;

public class Partition_list {
    public ListNode partition(ListNode head, int x) {
        // write your code here
        if(head == null){
            return null;
        }
        ListNode leftDummy = new ListNode(0);
        ListNode rightDymmy = new ListNode(0);

        ListNode left = leftDummy, right = rightDymmy;

        while(head != null){
            if(head.val < x){
                left.next = head;
                left = head;
            }
            else{
                right.next = head;
                right = head;
            }
            head = head.next;
        }
        right.next = null;
        left.next = rightDymmy.next;
        return leftDummy.next;
    }
}

File: python_version/linked-lists/partition_list.py
"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""

class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
class Solution:
    """
    @param head: The first node of linked list
    @param x: An integer
    @return: A ListNode
    """

    def partition(self, head, x):
        # write your code here
        if head is None:
            return head
        aHead, bHead = ListNode(0), ListNode(0)
        aTail, bTail = aHead, bHead

        while head is not None:
            if head.val < x:
                aTail.next = head
                aTail = aTail.next

            else:
                bTail.next = head
                bTail = bTail.next

            head = head.next

        bTail.next = None
        aTail.next = bHead.next

        return aHead.next